### ISSUE TITLE
Refactor XDS ACK handling

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#25587]
In latest builds something seems off with the EDS handling - I think its not always triggering pushes when it should. 

This is a very vague issue, will investigate and update

**Comment by howardjohn** [comment]
Couldn't reproduce, I think I had local changes breaking things

**Comment by howardjohn** [comment]
I can reproduce. The key here is things go weird when istiod is restarted. So basically just have a pod, then restart istiod. You will see:
```
2020-07-18T00:07:55.721706Z     info    ads     ADS:CDS: REQ sidecar~10.244.0.8~shell-759b46c77f-k4lnc.default~default.svc.cluster.local-1 version:2020-07-18T00:07:06Z/4
2020-07-18T00:07:55.728569Z     info    ads     CDS: PUSH for node:shell-759b46c77f-k4lnc.default clusters:18 services:5 version:2020-07-18T00:07:54Z/3
2020-07-18T00:07:55.728866Z     debug   ads     ADS:EDS: ACK sidecar~10.244.0.8~shell-759b46c77f-k4lnc.default~default.svc.cluster.local-1 2020-07-18T00:07:06Z/4 Qry7afpcxro=3d2092af-e88d-41ff-b507-856f7dd845b7
2020-07-18T00:07:55.729045Z     debug   ads     ADS:LDS: REQ sidecar~10.244.0.8~shell-759b46c77f-k4lnc.default~default.svc.cluster.local-1
2020-07-18T00:07:55.735973Z     info    ads     LDS: PUSH for node:shell-759b46c77f-k4lnc.default listeners:15
2020-07-18T00:07:55.736002Z     debug   ads     ADS:RDS: ACK sidecar~10.244.0.8~shell-759b46c77f-k4lnc.default~default.svc.cluster.local-1 2020-07-18T00:07:06Z/4 Qry7afpcxro=a80ce20f-bd58-42e9-b523-e38afd6962a7
2020-07-18T00:07:55.737920Z     debug   ads     ADS:EDS: REQ sidecar~10.244.0.8~shell-759b46c77f-k4lnc.default~default.svc.cluster.local-1 Expired nonce received Qry7afpcxro=3d2092af-e88d-41ff-b507-856f7dd845b7, sent
2020-07-18T00:07:55.744067Z     debug   ads     ADS:LDS: ACK sidecar~10.244.0.8~shell-759b46c77f-k4lnc.default~default.svc.cluster.local-1 2020-07-18T00:07:54Z/3 Qry7afpcxro=e1b12be9-62f3-4172-abaf-e38be7253337
2020-07-18T00:08:10.732870Z     debug   ads     ADS:EDS: REQ sidecar~10.244.0.8~shell-759b46c77f-k4lnc.default~default.svc.cluster.local-1 Expired nonce received Qry7afpcxro=3d2092af-e88d-41ff-b507-856f7dd845b7, sent
2020-07-18T00:08:10.733166Z     debug   ads     ADS:CDS: ACK sidecar~10.244.0.8~shell-759b46c77f-k4lnc.default~default.svc.cluster.local-1 2020-07-18T00:07:54Z/3 Qry7afpcxro=aed26953-985c-4814-9a5a-896ee2738c6c

```

The EDS and RDS requests are treated as ACKs when they should not be

**Comment by ramaraochavali** [comment]
I am guessing, envoy is sending empty resource names for the EDS request. Then only it can go in to that ACK mode because `listEqualUnordered` returns true (previous resource empty because Istiod is restarted and new resources empty - this should not happen). Some relevant issue https://github.com/envoyproxy/envoy/issues/7529. The conclusion there was to skip sending response. Not sure if it is same though - but I can not think of any other way this can happen.

https://github.com/istio/istio/pull/25629 - Sends response in that case. Can you please check if with this you are seeing EDS responses sent?

**Comment by costinm** [comment]
I have been thinking about this issue: and maybe we are just over-complicating things by focusing too much on ack/nonce craziness inherited from the non-aggregated protocols, when there is a much simpler solution.

Let's look at a typical pubsub system. A client sends a subscription for a topic, with some optional query attributes. Later on
the user can update the subscription, or change the attributes. The pubsub system will take each received subscription request and update so it sends only the required data.

In most cases, we can do the same thing for ADS: ignore version_info and nonce ( use them for tracking latencies or other things ), and just rely on the typeURL and the 'watched resource' list. 

We do need to keep track of the 'active subscriptions' - we do now for the new generators - so the logic should be 'if it's a new subscription or an update adding/changing the attributes'. 

In parallel, version_info provides info about the 'active' state of each resource, while nonce allow us to track latency. 

If we send LDS/CDS changes to Envoy that update the EDS/RDS list: envoy will have to send the update resource list, and the 
danger is that ACKs/NACKS for 'in flight' messages will be returned AFTER the 'update resource list'. The doc in envoy makes no guarantee that I found that each DiscovertRequest.resources reflect the current interest list - and is not a mirror of the 
received resources getting acked.  The proper solution would be to convince envoy to document/add tests  to clarify that for 
a typeURL, the 'resources' always reflect the current list.  Until that happens - we still need to identify this by tracking all send and un-acked messages, by nonce. 

Unfortunately, until Envoy clarifies the semantic of the resource[] in a delayed ACK it is still not clear if the delayed ACK is just an ACK and should have no impact of the watched resource list.

**Comment by howardjohn** [comment]
Drop from release blocker to P2 as the initial issue is fixed, will keep open for discussion about changing overall

**Comment by howardjohn** [comment]
Fixed by #28261
